test/integration/header_integration_test.cc

```diff
@@ -0,0 +1,454 @@
+#include "common/protobuf/protobuf.h"
+
+#include "test/integration/http_integration.h"
+
+#include "api/filter/http/http_connection_manager.pb.h"
+#include "gtest/gtest.h"
+
+namespace Envoy {
+
+class HeaderIntegrationTest : public HttpIntegrationTest,
+                              public testing::TestWithParam<Network::Address::IpVersion> {
+public:
+  HeaderIntegrationTest() : HttpIntegrationTest(Http::CodecClient::Type::HTTP1, GetParam()) {}
+
+  enum HeaderMode {
+    Append = 1,
+    Replace = 2,
+  };
+
+  void addHeader(Protobuf::RepeatedPtrField<envoy::api::v2::HeaderValueOption>* field,
+                 const std::string& key, const std::string& value, bool append) {
+    envoy::api::v2::HeaderValueOption* header_value_option = field->Add();
+    auto* mutable_header = header_value_option->mutable_header();
+    mutable_header->set_key(key);
+    mutable_header->set_value(value);
+    header_value_option->mutable_append()->set_value(append);
+  }
+
+  void initializeFilter(HeaderMode mode, bool includeRouteConfigHeaders) {
+    config_helper_.addConfigModifier([&](envoy::api::v2::filter::http::HttpConnectionManager& hcm) {
+      using namespace envoy::api::v2::filter::http;
+
+      const bool append = mode == HeaderMode::Append;
+
+      auto* route_config = hcm.mutable_route_config();
```

Comment:
nit: I find large proto constructions like this to be almost unreadable. Is there any ability to write in YAML and do a conversion? It's fine to say no. If no, can you add some comments on what the giant construction block is doing?